### DISCUSSION TITLE
Namespace System.Numerics.Generic - IEEE 754 and custom number types

### DISCUSSION THREAD

**Discussion by c-ohle** [dotnet/runtime#74507] [Ideas]
### I would like to propose a template solution to support IEEE 754 floating-point and decimal types in arbitrary size.

A implementation for floating-point already exists and can be tested: [Generic.Float.cs](https://github.com/c-ohle/RationalNumerics/blob/master/System.Numerics.Rational/Generic.Float.cs).
(It is currently part of a [project for arbitrary arithmetic](https://github.com/c-ohle/RationalNumerics) in general that can be compiled for NET6 and NET7.) 

The function set is identical to the system's other floating-point types and current NET7 naming conventions. `INumber` / `IBinaryFloatingPointIeee754` is fully supported.
For the user it is easy to use and to define number types in any desired precision: 

![image](https://user-images.githubusercontent.com/63112403/186301640-89fb463b-2fa9-4995-89bc-cf218cba8f4c.png)

Difference to the system types are only the constants implemented as static properties.

![image](https://user-images.githubusercontent.com/63112403/186305808-86829c7c-73b6-421f-a2df-c562e5caefe2.png)

It's not all about to have floating-point for higher precision. It is also to support old and new special formats like Extended [Float80](https://github.com/c-ohle/RationalNumerics/blob/65f22629e062ce467a13f3851a46ea6dcbee6f0e/System.Numerics.Rational/Generic.Types.cs#L308) or to emulate the processors with specific internal precisision, rounding modes etc. 
To make this easy the solution includes a `IFloat<T>` template that allows to implement new special number types based on `Float<T>` that inherit the base functionality inclusive all the `INumber` interfaces.
Examples for this can be found here: [Generic.Types.cs](https://github.com/c-ohle/RationalNumerics/blob/master/System.Numerics.Rational/Generic.Types.cs)

### How it works

The template type definition is simple as possible. 
The template parameter defines the struct size and the struct size the default IEEE 735 floating-point type and the binary layout:

```csharp
public readonly struct Float<T> where T : unmanaged
{
  readonly T p;
}
```
Based on this principle, a set of number type templates are possible, which could stay in a namespace `System.Numerics.Generic`:

![image](https://user-images.githubusercontent.com/63112403/186325528-72c8775d-19b9-4ee6-848f-1215b8b2f74e.png)

(The implementation for `Decimal<T>` is already in work.)

Internally, all template-based number types can use the same calculation core. Once initialized, this core does not require any further memory allocation at runtime. Since the number data is stored in the fixed size defined in the template, calculations with such number types can be performed without memory allocations. This is in contrast to other arbitrary number types such as BigInteger or BigRational, where dynamic buffers are required.

**Comment by tannergooding** [comment]
Reiterating what I said in https://github.com/dotnet/runtime/issues/74475, this is not something the BCL would be interested in providing.

Having a type that takes a `T` but which doesn't actually use the `T` for anything other than sizing information is problematic on multiple levels and I don't see such a proposal ever passing API review. It also comes with numerous usability concerns since a user could pass in `(byte, Guid, nuint)` and due to the `LayoutKind.Auto` mechanics of `ValueTuple` it may differ in size and behavior across different platforms/architectures.

Providing non-standard or largely deprecated formats, such as the 80-bit "extended-precision" format provided by the x87 FPU, is a non-goal of the BCL. Adding more "custom" formats is likewise problematic and comes with many negative downsides.

We may in the future provide an IEEE 754 `binary128` type (sometimes called `Quad`), but that would be an explicit `Float128` or similar name. We may likewise in the future provide some `BigFloat` type, but that would likewise follow the IEEE 754 requirements and would match existing BCL patterns to ensure that developers can easily use and intuitively understand the type and its API surface.

**Reply by c-ohle** [reply]
@tannergooding 
> Having a type that takes a T but which doesn't actually use the T for anything other than sizing information...

So far, this is for compatibility with NET 6, where we don't have a static interface solution. For NET 7 and later, we can constrain the template parameter and the interface itself can more explicitly define specific processor types, the internal precision rounding modes, and so on. The calculation core already use flags with default values for this.

> and due to the LayoutKind.Auto mechanics of ValueTuple it may differ

ValueTuple was only to show the principle with a known type - but in general:
The runtime size of `Float<T>` is not importand. Only for the template it self where `sizeof(T)` has to match.
Example: a dedicated implementation for Extended [Float80](https://github.com/c-ohle/RationalNumerics/blob/d1d2d7756d3871a762363673811fb7b160f9fd49/System.Numerics.Rational/Generic.Types.cs#L310) internal:
```csharp 
    [StructLayout(LayoutKind.Sequential, Size = 10)]
    private readonly struct UInt80 { }
    private readonly Float<UInt80> p;
```
This works on all platforms and the types can be binary compatible. 

> largely deprecated formats, such as the 80-bit "extended-precision" format provided by the x87 FPU, is a non-goal of the BCL

It's a template solution and Float80 only a example how to implement a dedicated number type based on this, not intended as part of the BCL. But it's also good test that shows that user can define in some lines a compatible a extended type that can be used for example  to read old binary file formates and convert to double.

> We may in the future provide an IEEE 754 binary128 type

Why not now when you can have this and much more? 
As template it cost not much memory and allows float128, float256, float512, ... tested up to float4096 the extended versions too and custom sizes like float96 or what ever precision is required.

> We may likewise in the future provide some BigFloat type

I would advise against that. 
What I have done before was to implement exactly this - such BigFloat type, dynamic, array based etc. 
It works but makes not much sense since the effective precise digits size is not in 
any reasonable proportion to the data size to manage.
But that led me to the template solution and is tested finally also more performant.

**Reply by tannergooding** [reply]
That is not an approach we'd expose in the BCL, it is not something that would pass API review.

The approaches and potential future directions I have described here and in other issues have very explicit reasons why they get chosen and will continue to be chosen for future types or APIs exposed by the BCL.

**Reply by c-ohle** [reply]
@tannergooding It's a discussion. I would really like to know which criteria are violated.

* The user interface? The template types are 100% compatible with the other system floating-point types. Every name, every function, property and operator, the standard interfaces, string formatting, parsing, all the same. Binary compatible too for the same IEEE formats. I suppose that this can not be the problem.
* Is it because it based on templates? We have templates for so many similar cases like collections and expressions. It makes always sense if types are rarely used but if then in many possible configurations.

**Reply by bartonjs** [reply]
The part that violates guidelines (or, rather, is so strange that it doesn't have a guideline explicitly against it) is that the type is generic, but the `T` doesn't appear in the API anywhere, it's just a size value.

**Reply by c-ohle** [reply]
@bartonjs thanks for discussion. 

As mentioned before, so far it was for compatibility with NET6. For NET7 and later the plan is to constrain the type parameter with a static interface what allows also more specific informations for processor emulations etc. The information needed for the algorythms matches in a 64 bit field. Such interface based floting-point type definition can therefore be a really small and compact but readable. 
It's currently [as static fields](https://github.com/c-ohle/RationalNumerics/blob/d1d2d7756d3871a762363673811fb7b160f9fd49/System.Numerics.Rational/Generic.Float.cs#L736) but the static props below can show how the interface could look like.

**Reply by tannergooding** [reply]
> It's a discussion.

Its a discussion that was stemmed off of an issue where the area owners had already explicitly stated "no" and why.

> I would really like to know which criteria are violated.

There have been multiple examples given in the original issue, which were reiterated in this discussion, and similar examples given in some of the other issues opened.

> so far it was for compatibility with NET6. For NET7 and later the plan is to constrain the type parameter with a static interface what allows also more specific informations for processor emulations etc.

You cannot make a generic constraint more restrictive in a later version of the code, such changes are breaking and that's a "core" consideration for anything exposed in the BCL (stability over time to ensure binary breaks don't occur).

There is no scenario under which you could declare some `Float<long>` and have it map to a `double` or some `Float<UInt128>` and have it map to a hypothetical `Float128`. This is unintuitive and confusing behavior. It is likewise not infinitely extensible and becomes increasingly more verbose to use the more "bits" you want to represent (such as if you wanted to represent a `Float256` or larger).

--------------------------

More generally, the BCL follows the "Framework Design Guidelines". Most of the rules for which can be found here: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/

We then have an entire book that expands on this in more detail including with annotations from members of the API review team: https://www.amazon.com/Framework-Design-Guidelines-Conventions-Addison-Wesley/dp/0135896460/

Almost anything exposed by the BCL is available to all developers "out of the box". We have very stringent requirements to maintain backwards compatibility, to minimally consider forward compatibility so logical extensions to a given type/API can still be made, to be generally consistent with what is already exposed by the BCL, to ensure that types/APIs are easily usable by the target audience, to ensure that code is intuitive to read/maintain/understand, etc.

As indicated the proposals provided so far do not meet this criteria. The area owners are interested in providing some features like `Rational`, 128-bit IEEE 754 binary float, IEEE 754 `decimal`, or even "big float/decimal". However, such APIs will need to fit the existing shapes, standards/conventions, and expectations that existing and similar BCL types already follow.

**Reply by c-ohle** [reply]
@tannergooding 
> stated "no" and why.

There was no remaining argument why. 

> There have been multiple examples given in the original issue

It's a whole new approach - unlike the BigRational discussion we had and the technical issues there.

> You cannot make a generic constraint more restrictive in a later version of the code

By later I mean tomorrow and not a later framework version. 

> There is no scenario under which you could declare some Float<long> ...

it would look like this:

```csharp

  Float<FloatType128X64> x;  
  Float<FloatType32SIMD> y;

```
or this:

```csharp

using Float128 = System.Numerics.Generic.Float<FloatType128X64>;
 
```
Where FloatType128X64 is like a really small key struct type. 
Nothing confusing. Everything is type safe and extensible based on the bitmask that already exists.

The real question is - is there a better solution to realize something like this?

**Reply by tannergooding** [reply]
> The real question is - is there a better solution to realize something like this?

"Better" is subjective and what is better for the BCL may not be what is best for your personal scenario or interests.

For the BCL, explicit types such as `Float128` or `Float256` are better than some generic type whose size is keyed off a "magic" `FloatType32SIMD`.

> Nothing confusing. Everything is type safe and extensible based on the bitmask that already exists.

Not confusing to you. This will be confusing to customers and other developers.

**Reply by c-ohle** [reply]
@tannergooding 
> For the BCL, explicit types such as Float128 or Float256 are better than some generic type whose size is keyed off a "magic" FloatType32SIMD.

And that's why I added already a Float128 type as example. **This is not a template**. 
For the user from outside is no difference to other float-number types. Value struct, fixed size, same  interfaces, same functions.
And full IBinaryFloatingPointIeee754 support. `IFloat<Float128>` serves only as implementation helper for this. 
Also possible to write all new. Makes finally no difference but more code. 
[full implementation](https://github.com/c-ohle/RationalNumerics/blob/master/System.Numerics.Rational/Generic.Types.cs)

```csharp
public readonly struct Float128 : IFloat<Float128>, IComparable<Float128>, IEquatable<Float128>, IComparable, ISpanFormattable    
{
    public static Float128 operator +(Float128 a, Float128 b) => new Float128(a.p + b.p);
    public static Float128 operator -(Float128 a, Float128 b) => new Float128(a.p - b.p);
    // and so on...

    //private:
    [StructLayout(LayoutKind.Sequential, Size = 16)]
    private readonly struct UInt128 { } // NET6 has no UInt128 
    private readonly Float<UInt128> p;
}
```  
It's all there what you want.
[End user view - documentation](https://c-ohle.github.io/RationalNumerics/api/System.Numerics.Float128.html)
